File: leetcode/cpp/274.h-index.cpp
class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n = citations.size();
        vector<int> counts(n+1, 0);
        for(int c: citations) {
            if (c > n) {
                c = n;
            }
            counts[c]++;
        }
        for (int i=n-1; i>-1; i--) {
            counts[i] += counts[i+1];
        }
        for (int c=n; c > -1; c--) {
            if (counts[c] >= c) {
                return c;
            }
        }
        return 0;
    }
};

File: leetcode/go/274.h-index.go
func hIndex(citations []int) int {
    sort.Sort(sort.Reverse(sort.IntSlice(citations)))
    h := 0
    for _, c := range citations {
        if c <= h {
            break;
        }
        h++
    }
    return h
}

/* */

func hIndex(citations []int) int {
  n := len(citations)
  dp := make([]int, n+1)
  for i := 0; i < n; i++ {
    cur := citations[i]
    if cur > n {
      cur = n
    }
    for cur > 0 {
      dp[cur]++
      cur--
    }
  }
  for i := n; i >= 0; i-- {
    if dp[i] >= i {
      return i
    }
  }
  return 0
}

